solo-io/gloo
# Declarative test setup

The declarative environment builder tool can setup Kubernetes test cluster(s) with IstioOperators, Helm Charts, and any Applications required for testing. This is based on the declarative setup tool used in GME.

### How do I get started? 

A make target is provided to run the setup. A config file needs to be provided via `CONFIG` variable.

To test the simple gloo gateway setup:

```bash
CONFIG=test/setup/example_configs/gloo-gateway-setup.yaml make setup-declarative-env
```

This can also be run directly with: 
```bash
go run ./test/setup setup -c test/setup/example_configs/gloo-gateway-setup.yaml
```
Make sure the images and helm chart are present before running the go command directly.

### What happens if I already have a (partial) setup? 

Running the declarative setup will not automatically fail if you have different resources already setup. This is slightly different from the regression tests which require you to set `SKIP_TEARDOWN` or `SKIP_INSTALL`.

For example:
1. If you have a kind cluster already running, the declarative setup will use the existing kind cluster
2. If you already have a namespace, the declarative setup will use the existing namespace, add any labels defined in the config file and continue. 
3. If you have gloo already installed, the declarative setup will run a `helm upgrade -i` and upgrade the existing installation. 

### How do I write my own config? 


Configuration is typically provided via yaml file.  Usage is usually coupled with `make` targets
since a lot of default environment variables are defined there.


The configuration file is broken into a list of clusters, each with a distinct set of capabilities:

```yaml
clusters:
    - name: management-cluster
      management: true
      kindConfig:
        ...
      istioOperators:
        ...
      charts:
        ...
      namespaces:
        ...
      apps:
        ...
      images:
        ...
```

* kindConfig: The configuration for the KinD cluster
* istioOperators: Installation mechanism for Istio uses istio operators. This is a list of istio operator configs to install into the cluster. Uses the istioctl set by the `ISTIO_VERSION` environment variable, or the installed istioctl binary if not set for installation.
* charts: helm charts to install into the cluster.
    * name: name of the chart.
    * namespace: namespace to install the chart into
    * path: if true, the chart is installed from the local filesystem. Otherwise, it is installed from the configured helm repo.
    * values: values to pass to the chart.
* namespaces: namespaces to create in the cluster. Labels can be specified on the namespace. 
* apps: applications to install into the cluster. Follows the k8s resources apis and includes deployments, serviceaccounts, services, configmaps, etc. 
* images: images to load into the cluster. If an image override is defined on the helm chart, the setup command will also load that into the kind cluster so images only need to be defined in one place.
